lib/unix.bzl

```
@@ -0,0 +1,326 @@
+        paths = cmd_paths,
+    )]
+
+unix_toolchain = rule(
+    attrs = {
+        cmd: attr.string(
+            doc = "Absolute path to the {} command.".format(cmd),
+            mandatory = False,
+        )
+        for cmd in _commands
+    },
+    doc = """
+A toolchain capturing standard Unix shell commands.
+
+Provides:
+  commands: Dict of String, an item per command holding the path to the executable or None.
+  paths: List of String, deduplicated bindir paths. Suitable for generating `$PATH`.
+
+Use `unix_configure` to create an instance of this toolchain.
+See `unix_make_variables` on how to use this toolchain in genrules.
+""",
+    implementation = _unix_toolchain_impl,
+)
+
+def _unix_make_variables_impl(ctx):
+    toolchain = ctx.toolchains[TOOLCHAIN_TYPE]
+    cmd_vars = {
+        "UNIX_%s" % cmd.upper(): cmd_path
+        for cmd in _commands
+        for cmd_path in [toolchain.commands[cmd]]
+        if cmd_path
+    }
+    return [platform_common.TemplateVariableInfo(cmd_vars)]
+
+unix_make_variables = rule(
+    doc = """
+Provides Unix toolchain commands as custom make variables.
+
+Make variables:
+  Provides a make variable of the form `UNIX_<COMMAND>` for each available
+  command, where `<COMMAND>` is the name of the command in upper case.
+
+Use `unix.MAKE_VARIABLES` instead of instantiating this rule yourself.
+
+Example:
+  >>> genrule(
+          name = "use-grep",
+          srcs = [":some-file"],
+          outs = ["grep-out"],
+          cmd = "$(UNIX_GREP) search $(execpath :some-file) > $(OUTS)",
+          toolchains = [unix.MAKE_VARIABLES],
+      )
+""",
+    implementation = _unix_make_variables_impl,
+    toolchains = [TOOLCHAIN_TYPE],
+)
+
+def _unix_config_impl(repository_ctx):
+    cpu = get_cpu_value(repository_ctx)
+    commands = {}
+    for cmd in _commands:
+        cmd_path = repository_ctx.which(cmd)
+        if cmd_path == None and cpu == "x64_windows":
+            cmd_path = repository_ctx.which(cmd + ".exe")
```

If a genrule uses `$(UNIX_FOO) -out=$@` but `foo` is missing, then I guess the command will be just `-out=<path>`. It won't be obvious what exactly failed.

Could you use a fake path for nonexistent tools? Like `foo-does-not-exist`, so the command fails as `foo-does-not-exist -out=<path>`.

---

If `foo` is missing, then no make variable will be generated in the first place (see [here](https://github.com/bazelbuild/bazel-skylib/pull/208/files#diff-31822a9b4af4c9d5c8b2163235dbe860R239)). So, in that case this will be a build failure with Bazel complaining that `$(UNIX_FOO)` is undefined. I think that's preferable to a fake path, as those could be silently ignored and cause dormant issues, depending on how they're used.

---

True! Thanks for clarifying.